### ISSUE TITLE
cdc-acm printing error

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#1915]
**_Reported by Dino Tinitigan:_**

Board: Arduino 101

When printing the same string pattern repeatedly in a loop, it occasionally fails and reprints the first 4 bytes of that string.

For example:

{code:java}
char *banner3 = "0.47\r\n";
while(1)
{
	write_data(dev, banner3, strlen(banner3));
}
```

will give an output similar to the attached screenshot

Similarly printing "ee*ee\r\n" repeatedly
outputs something like:
eee*ee
eee*ee
eee*eee*ee

Also printing the string "1\r\n"
will not show the issue because it is less then 4 bytes.


However if I were to print different string patterns I do not see the issue happen at all.

Attached is a the modified cdc-acm example the reproduces the issue and a screenshot of the terminal

I am able to reproduce the issue in Zephyr 1.6 and in Zephyr 1.7
I have tried adding delays between the prints yet I still see the issue

(Imported from Jira ZEP-2074)

**Comment by nashif** [comment]
**_by Dino Tinitigan:_**

* [cdcserialerror.PNG](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/cdcserialerror.PNG) (size: 6986)

**Comment by nashif** [comment]
**_by Dino Tinitigan:_**

* [cdc_acm.zip](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/cdc_acm.zip) (size: 1808373)

**Comment by nashif** [comment]
**_by Mark Linkmeyer:_**

I'm assigning this bug to _manoj subbarao_  so he can get it assigned to the right person on his team.

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

I was able to reproduce this issue today after running system for ~20-25 minutes with latest Code base. 

Here are my observations :-
1. In all the places where we got multiple characters printed, we got 2 times of same character set.
For example if we were printing 0.47, then it was printing 0.470.47. So looks buffer is getting printed 2 times.
2. I During analysis i observed that we are prininting two times in cases where "." is there in word. For example we are getting issue in 0.f1, 0.f2, 0.47, 0.48 nut we are not getting double character print in cases ee*eee like that.

I will continue debug this issue tomorrow and put findings on Jira.

**Comment by nashif** [comment]
**_by Andy Ross:_**

With the initial caveat that I'm looking at this driver for the first time:

This looks like an application bug in our cdc_acm sample to me.  There are two:

The first is that the code is expecting uart_irq_tx_ready() indicates that the complete buffer passed to uart_fifo_fill() has been transmitted.  It doesn't, it's just an indication that the TX pipe can accept at least one byte (strictly it means that the 8250-style interrupt "has been raised", which means the same thing).  On traditional UART hardware, the code as written would never work without overflow.  *But*, on top of the cdc_acm driver I think this is benign.  That UART simply forwards the data down into usb_write() on the IN endpoint, and lets the buffer handling happen there.  When the endpoint callback happens out of the USB stack, we know that all the buffered data has been transmitted.

The second bug is the real culprit I think: the code isn't checking the return value of usb_fifo_fill(), which returns the number of bytes consumed from the buffer.  Now, most of the time it seems that USB pipes are really fast[1] and can keep up with the MCU, but occasionally we get an buffer full condition and usb_fifo_fill() can't queue all the bytes in usb_write() and returns a short cound.  And those bytes get dropped, because the application ignores the issue.

[1] Interestingly, I don't see any code in the cdc_acm driver to actually implement baud rate throttling.  It handles the protocol to allow the other side to "set" the speed, but then it will just spit bytes back to the user as fast as the application code provides them.

**Comment by nashif** [comment]
**_by Andy Ross:_**

Short version: can someone with a test rig ready try the following (*completely untested!*) version of write_data() and see if it fixes the problem?

```
static void write_data(struct device *dev, const char *buf, int len)
{
	int sent;

	uart_irq_tx_enable(dev);

	while (len) {
		data_transmitted = false;
		sent = uart_fifo_fill(dev, (const u8_t *)buf, len);
		while (data_transmitted == false)
			;

		len -= sent;
		buf += sent;
	}

	uart_irq_tx_disable(dev);
}
```

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

Tried above patch from Andy, Issue still persists.



Our observations from today's Debug :- 
Debug Method: Sprinkled Prints in the application/driver code. (sample debug patch attached)

Observation 1:- [Erroneous Pattern identified]
    1 out of ~3000 data write into USB EP,  first 4 data bytes sent to USB endpoint are getting printed twice  on ttyACM0– which is  an issue
For example 
A.	If we  write “abcdef” into USB Endpoint, in erroneous case it prints “abcdabcdef
B.	If we are writing “0.40” into USB Endpoint, it is printing “0.400.40

Observation 2:- [USB Endpoint data seems to be proper even when we get error print in ttyACM0]  
We enabled prints to debug in the  USB Driver and found that correct data is written into the USB Endpoint. 
For Example: Even in case when ttyACM0 showed error pattern  “abcdabcdef” , correct pattern of “abcdef” was written into USB endpoint.

Tried Solutions
----------------------

Sol 1 :- (Adding Delay between packets)
Tried with k_sleep() / k_busy_wait()  between 2 data transmit to give enough settling time.
 Issue is still seen.

Sol2:-  (Flushing of Endpoint data buffer)
Tried  flushing the endpoint  data  buffer to make sure the stale data is not picked up  between 2 consecutive packet send. (calling usb_dc_ep_flush() ).
But issue still persists.

Need to debug more.

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

Further update on this issue :-

Experiments 1 :- Issue still visible on changing endpoint. 
We change Input Endpoint from Endpoint 4 to Endpoint 5, Endpoint 3, Endpoint 2. But issue is still visible. So this indicates that issue is visible on all Endpoints.

Experiment 2:- 
we captured data using usbmon at Host side (Linux Machine where we are sending data via ttyACM0), 
when issue is occurring, it indicates that we are writing 8 bytes into USB Endpoint but we are receiving 12 bytes on linux host side.

At erogenous case Data written into USB Endpoint  (abcdef\r\n):-

**2782 Youve_len 8 datalen 8 and Data abcdef

** 0 ** 97 98 99 100    ** 4 ** 101 102 13 10        <---- 8 bytes written (abcdef\r\n)
=interrupt_handler===> 2782 ===

Data received on USB Host(Linux Machine where ttyACM0 is connected).  Here is output when we are getting 12 bytes on Host Machine. (61626364 61626364 65660d0a ==> abcd abcd ef\n\r)
ffff8e5f47370600 386720162 S Bi:3:002:1 -115 512 <
ffff8e5f3a52a0c0 386720223 S Ci:3:002:0 s c0 05 0000 0000 0002 2 <
ffff8e5f46775b40 386720306 C Bi:3:004:4 0 12 = 61626364 61626364 65660d0a   -<---- 12 bytes (abcdabcdef\r\n)

Outcome:- Even though we are writing 2 bytes but endpoint is sending 12 bytes. Looks like something crazy is happening.


Experiment 3:-
We encapsulated all Endpoint registor write with irq_lock() and irq_unlock(), issue is still reproducible.


Experiment 4:- We verified the value written into control regs is same in error case and non error case
control register (diepctl) = 0x800988080
packet/data length register (dieptsiz) = 80008 //single packet with 8 bytes

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

* [USB_Analyser.png](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/USB_Analyser.png) (size: 93325)

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

I captured data transferred through USB protocol analyzer and it indicates that in error case it is is receiving 12 bytes(instead of 8 bytes),
Even we wrote only 8 bytes but it received 12 bytes, I have attached snapshot. !USB_Analyser.png|thumbnail!

**Comment by nashif** [comment]
**_by Andy Ross:_**

* [cdc_acm_buffering.patch](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/cdc_acm_buffering.patch) (size: 3439)

**Comment by nashif** [comment]
**_by Andy Ross:_**

I spent some time playing with this today.  The root cause, as Youvedeep discovered, seems to be in the hardware.  Our usage isn't incorrect per documentation, and in fact matches the FIFO behavior of all the other dwc2 drivers I found.

I also see a different issue not yet discussed here: the USB driver will occasionally hang, usually soon after startup.  This seems to coincide (usually, but not always) with control transfers (endpoint zero) happening simultaneous to the transmitted IN endpoint data.  When this happens (again, usually), I see debugging logging I added fire, and it indicates that EP0 sees that its FIFO is not empty as expected but instead is short by a number of bytes equal to one of our ACM packets!  So it seems like something is mixing up the FIFOs internally?  The hardware does have a "shared FIFO" mode that can be synthesized, but the Intel Quark SE datasheet says pretty clearly that it's not enabled and all endpoints have dedicated FIFOs.  This may need some hardware people to chime in (not here on Jira, obviously).

But regardless: at least we know that the duplicated-FIFO-words problem becomes much rarer when the packets are large.  And our cdc_acm UART implementation does synchronous writes and thus is really bad at buffering big packets.

So I wrote a quick implementation of buffering in cdc_acm which will transmit a small packet if the hardware is idle, but otherwise keeps a 64-byte (i.e. one full speed USB packet) buffer to coalesce while the hardware is working.  With this, and a write_data() fix in the sample (inline below -- this one is tested and works).  With this, I get one error only every megabyte or so of data transfer, which is at a level where most serial protocols can tolerate.  Maybe this will be good enough as a workaround.

I'll point out that it's also *much* faster than the current code for streams of small packets -- I can get almost 2Mbps down the pipe instead of the 80-100kbps we were seeing earlier.

I'll attach the cdc_acm.c patch as a file here to avoid teething trouble with github.  Will have it cleaned up for submission tomorrow.

This is what you need to fix in write_data:

```
static void write_data(struct device *dev, const char *buf, int len)
{
	int sz;

	uart_irq_tx_enable(dev);

	while (len > 0) {
		data_transmitted = false;
		sz = uart_fifo_fill(dev, (const u8_t *)buf, len);
		if (sz != len) {
			while (data_transmitted == false)
				k_busy_wait(50);
		}
		len -= sz;
		buf += sz;
	}

	uart_irq_tx_disable(dev);
}
```
 [^cdc_acm_buffering.patch]

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

* [ACM0_USB](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/ACM0_USB) (size: 30520)

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

* [UART_Reg_Dump](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/UART_Reg_Dump) (size: 4825257)

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

I captured status of USB register status. 
ACM0_USB  : Data sent over cdc_acm
UART_Reg_Dump : IA Log on UART
Please see that we are encountering error in 4075 th time we are sending data.

Corresponding to 4075th iteration print on UART console shows the status of USB registers :-
** 4075   0 ** 112 113 114 115  ** 4075   4 ** 116 117 13 10    USB registers: for 4075
  EP 0 registers:    DIEPCTL : 0x80008000           DIEPINT : 0x20d0        DIEPTSIZ: 0x180000      DIEPDMA : 0xa8000aa4
  EP 1 registers:    DIEPCTL : 0xc8010      DIEPINT : 0x2090        DIEPTSIZ: 0x60000000            DIEPDMA : 0x0
  EP 2 registers:    DIEPCTL : 0x0          DIEPINT : 0x80          DIEPTSIZ: 0x0           DIEPDMA : 0x0
  EP 3 registers:    DIEPCTL : 0x0          DIEPINT : 0x80          DIEPTSIZ: 0x0           DIEPDMA : 0x0
  EP 4 registers:    DIEPCTL : 0x98040      DIEPINT : 0x2090        DIEPTSIZ: 0x0           DIEPDMA : 0x0

  EP 5 registers:    DIEPCTL : 0x0          DIEPINT : 0x80          DIEPTSIZ: 0x0           DIEPDMA : 0x0
*****USV Regs****
GOTGCTL : 0xd0000       
GOTGINT : 0x0           
GAHBCFG : 0x1           
GUSBCFG : 0x1440
GINTSTS : 0x54008428    
GINTMSK : 0x800c3810    
DCFG    : 0x8200173     
DCTL    : 0x0
DSTS    : 0x453f06      
DIEPMSK : 0x1           
DOEPMSK : 0x0           
DAINT   : 0x0
DAINTMSK: 0x90013       
GHWCFG1 : 0x500         
GHWCFG2 : 0x2280d514    
GHWCFG3 : 0x1240468     
GHWCFG4 : 0x16008020
 Aaialbe space issue - 34 and avail - 136

This iteration shows that we have writeen only 8 bytes on USB endpoint but we are seeing 3 bytes on host side (ttyACM0).


Also the status of registers in between good case (pqrstu) and error case (pqrspqrstu) is only in 2 places :-
DTS Resiter : This register is different as frame number is changing with each packet.
IP_EP[4].DIEPTCL :- The difference is observed in 16th bit (it is toggling between 0 and 1). in Datasheet 16th bit is shown as reserved bit.
Also But this does not seem to be issue as it is toggeling on alternative frames even in good case.
So this does not seem to be a issue.


Also i confirmed few more things :-
1. We are always sending data in slave mode.
2. Maximum Packet size is 64 bytes.
3. Maximum number of packets which can be sent is 2.
4. Clock freq is 48 MHz.
5. LP Mode is disabled. (GHWCFG3). [^ACM0_USB]  [^UART_Reg_Dump]

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

* [YOUVE_USBC](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/YOUVE_USBC) (size: 572545)

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

* [youve_UARTC](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/youve_UARTC) (size: 6928478)

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

* [Screenshot from 2017-05-05 16-19-18.png](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-1915/Screenshot from 2017-05-05 16-19-18.png) (size: 357884)

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

I did one simple experiment which indicates that Data written by us on Endpoint is getting written only once even in Error Case.
DTXFSTSi ==> shows the space available on IP Endpoint Buffer i (in terms of u32).  

Experiment Description :-
I read DTXFSTS4 register just before and after the putting Data into Endpoint register. As expected the difference between space available before and After putting Data should be equal to number of bytes written in Endpoint.

Data written on Endpoint :- pqrstu\r\n.

Good case:- 
if we are writing 8 bytes (pqrstu\r\n)  into endpoint buffer, it indicates that before putting Data in Endpoint there was 136 bytes available. Then we put 8 bytes into endpoint, this resulted in 128 bytes into endpoint (136 - 8).
=before==136=   ** 2109  0 ** 112 113 114 115   ** 2109  4 ** 116 117 13 10     =After==128==

Error Case :-
On Error condition at (pqrspqrstu\r\n) (please look into attached file YOUVE_USBC) in  2113th iteration of Data transfer we can see that Data is written on Endpoint only once. But Host machine has received 12 bytes.
When i looked at endpoint side,  Before writing Data there was 136 bytes available and after writing Data there were 128 bytes space available.
So this Experiment indicates that we are Just writing 8 bytes.
=before==136=   ** 2109  0 ** 112 113 114 115   ** 2109  4 ** 116 117 13 10     =After==128==


Here is simple snapshot of my debug prints :-
   key = irq_lock();
+      printf("=before==%d=\t",USB_DW->in_ep_reg[4].dtxfsts*4);
        for (i = 0; i < data_len; i += 4) {
 +               printf("** %d  %d ** %u %u %u %u \t", youve_count, i, data[0+i], data[1+i], data[2+i], data[3+i]);
                USB_DW_EP_FIFO(ep_idx) = *(volatile u32_t *)(data + i);
        }
+     printf("=After==%d==\n",USB_DW->in_ep_reg[4].dtxfsts*4);
        irq_unlock(key);
*youve_count is getting incremented every time data is written.


File Description :- 
youve_UARTC :-- > Log Dump on UART Console (ttyUSB0).
YOUVE_USBC : Data Transferred to Host Machine Via USB (ttyACM0)
*.png :- Image showing comparison of Data sent Vs Data received.
 [^youve_UARTC]  [^YOUVE_USBC]  !Screenshot from 2017-05-05 16-19-18.png|thumbnail!

**Comment by nashif** [comment]
**_by Youvedeep Singh:_**

Few more observation :-
1. If we are transfering less than or equal to 4 bytes we are not getting duplication of first 4 bytes. For example - if we are transfering "abcd" or "ab\r\n" we did not even single duplication of data in 12 hours of run (over night).

2. It is only first 4 byte which are getting dumpicated and rest bytes are never getting dumpicated.
For Example : - 
abc\r\n --> abc\rabc\r\n---- which looks like abcabc
abcdef ---> abcdabcdef
abcdefgh --> abcdabcdefgh

---> Looking at this behavior, it looks like  "When endpoint starts sending 2nd dword(4 bytes), somehow pointer is moving back to first 4 bytes (First dword) and it sends first 4 bytes again."



One more experiments tried :-
1. We changed  "Input end point buffer depth" from 0x22 to 0x10 (minimum) but it did not changed any behavior.
